src/byte_sniffer.rs

```diff
@@ -0,0 +1,138 @@
+use futures_util::io::AsyncRead;
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+#[pin_project]
+pub(crate) struct ByteSniffer<Reader>
+where
+    Reader: AsyncRead,
+{
+    /// The wrapped [`AsyncRead`]er
+    #[pin]
+    inner: Reader,
+
+    /// Current length accumulator
+    current_length: usize,
+
+    /// The configured maximum length
+    max_length: usize,
+
+    upload_clamped: Option<Arc<AtomicBool>>,
+}
+
+impl<Reader> ByteSniffer<Reader>
+where
+    Reader: AsyncRead,
+{
+    pub fn new(max_length: usize, inner: Reader) -> Self {
+        let current_length = 0;
+
+        Self {
+            inner,
+            current_length,
+            max_length,
+            upload_clamped: None,
+        }
+    }
+
+    pub fn with_callback(mut self, flag: Arc<AtomicBool>) -> Self {
```

Comment:
Only that using a callback in this way is fairly tightly coupled to what is otherwise a pretty generic module. If I ever published this as a separate crate I would to be forcing consumers to pass a no-op closure for no reason